design/batch_retrieval_design.md

```diff
@@ -1,20 +1,176 @@
+| 2    | *Server side:*<br />Update Batch retrieval to return list of variables **and** their response (success/failure) (**) | - Will help us during rotation for Milestone 2<br />- Best practice / straightforward design for batch endpoint | - Requires both client/server implementation<br /> | M, 10 days            |
+| 3    | Stay as is                                                   | No additional work needed                                    | - Bad UX<br />- Performance issues                 | -                     |
 
 (*) Fallback solution: use this solution as a safety net, only if the original Batch Retrieval request fails
 
 (**) This solution can be broken up in two: 1. Create a new API endpoint or 2. use existing one (return 409 response)
 
+*Decision:* We decided to go with #2 and create our own API. We decided to create a new batch retrieval API because of the following reasons:
+
+1. Adding onto the existing batch endpoint would lead to changing behaviors for our needs in the Secrets Provider which would have cascading impacts on other teams. 
+2. The two APIs (current and new) offer separate use-cases;  partial success (new) and fail fast (old)
+3. The old batch retrieval API accepts variableIDs as query parameters. We anticipate a long string of variableIDs to be passed in for secret retrieval and this impacts performance.
+
+ The new batch retrieval API endpoint will not be forward facing and be for **internal use only**. 
+
+### Design
+
+#### Secrets Provider Job flow
+
+As a small recap, the following explains the Secrets Provider Job flow
+
+1. The `cyberark-secrets-provider-for-k8s` runs as a Job and authenticates with the Conjur/DAP Follower using the Kubernetes Authenticator (`authn-k8s`).
+
+2. The `cyberark-secrets-provider-for-k8s` reads all Kubernetes secrets required by the applications in the same namespace.
+
+3. For each mapped Kubernetes secret, 
+
+   a. The `cyberark-secrets-provider-for-k8s` performs a Batch Retrieval request and retrieves Conjur/DAP secrets.
+
+   b. It updates the Kubernetes secret with the Conjur/DAP secret value.
+
+4. The `cyberark-secrets-provider-for-k8s` Job is completed.
+
+The new Batch Retrieval API will effect #2a. 
+
+### URI
+
+##### POST /secrets/batch
+
+This method retuns a list of secrets, their status code, and their value if an error did not occur. See difference between proposed API and existing API see [Proposed API vs Old API](#proposed-api-vs-old-api) below.
+
+```
+POST /secrets/batch
+
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+### Request Body
+
+The request body is a list of all variable IDs (paths to secrets in Conjur/DAP) that need to be retrieved. For example: 
+
+```
+{
+  "variables": [
```

Comment:
It needs to be variables here no? because in conjur secrets are referenced to as variables

Reply:
I think I got the definition of `variables` in Conjur.
variables are Conjur entities which hold secrets as their value.
They may contain other info such as attributes etc.
So, `secrets` API should return only the variables **values**, as correctly done by the existing `secrets` API.
But this proposed new API returns more info than just the values, so maybe it should not be called `secrets`.
Maybe it should be variables, because it can bring more info on the variables besides the secret (which is the variable value).